File: qa-portal/backend/app/schemas/automated.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime
from uuid import UUID
from app.models.automated import AutomatedRunStatus, AutomatedResultStatus
from app.schemas.user import UserSummary


# ─── Suite ────────────────────────────────────────────────────────────────────

class SuiteCreate(BaseModel):
    project_id: UUID
    name: str
    description: Optional[str] = None


class SuiteUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None


class SuiteResponse(BaseModel):
    id: UUID
    project_id: UUID
    name: str
    description: Optional[str]
    created_by: UUID
    creator: UserSummary
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


class SuiteSummary(BaseModel):
    id: UUID
    project_id: UUID
    name: str
    description: Optional[str]
    created_at: datetime
    test_count: int = 0
    last_run_status: Optional[str] = None
    last_run_at: Optional[datetime] = None
    last_run_passed: int = 0
    last_run_total: int = 0



# ─── Automated Test Case ─────────────────────────────────────────────────────

class AutoTestCreate(BaseModel):
    name: str
    description: Optional[str] = None
    source_test_case_id: Optional[UUID] = None
    script_code: Optional[str] = None
    target_url: Optional[str] = None
    order: Optional[int] = 0


class AutoTestUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None
    script_code: Optional[str] = None
    target_url: Optional[str] = None
    order: Optional[int] = None
    is_active: Optional[bool] = None


class AutoTestResponse(BaseModel):
    id: UUID
    suite_id: UUID
    name: str
    description: Optional[str]
    source_test_case_id: Optional[UUID]
    script_code: Optional[str]
    target_url: Optional[str]
    order: int
    is_active: bool
    created_by: UUID
    created_at: datetime
    updated_at: datetime
    model_config = {"from_attributes": True}


# ─── Clone from manual ───────────────────────────────────────────────────────

class CloneFromManualRequest(BaseModel):
    suite_id: UUID
    source_test_case_id: UUID
    target_url: str


# ─── IA Script Generation ────────────────────────────────────────────────────

class GenerateScriptRequest(BaseModel):
    description: str
    target_url: str
    source_steps: Optional[List[dict]] = None


class GenerateScriptResponse(BaseModel):
    script_code: str


class RefineScriptRequest(BaseModel):
    current_script: str
    instruction: str


class RefineScriptResponse(BaseModel):
    script_code: str



# ─── Run ──────────────────────────────────────────────────────────────────────

class RunCreate(BaseModel):
    environment: Optional[str] = None
    version: Optional[str] = None


class RunResponse(BaseModel):
    id: UUID
    suite_id: UUID
    status: AutomatedRunStatus
    environment: Optional[str]
    version: Optional[str]
    started_at: datetime
    finished_at: Optional[datetime]
    triggered_by: UUID
    model_config = {"from_attributes": True}


class RunSummary(BaseModel):
    id: UUID
    suite_id: UUID
    status: AutomatedRunStatus
    environment: Optional[str]
    version: Optional[str]
    started_at: datetime
    finished_at: Optional[datetime]
    total: int = 0
    passed: int = 0
    failed: int = 0
    error: int = 0
    skipped: int = 0


class RunStatusResponse(BaseModel):
    id: UUID
    status: AutomatedRunStatus
    total: int = 0
    completed: int = 0
    passed: int = 0
    failed: int = 0
    error: int = 0


# ─── Run Result ───────────────────────────────────────────────────────────────

class RunResultResponse(BaseModel):
    id: UUID
    run_id: UUID
    automated_test_case_id: UUID
    test_name: str = ""
    status: AutomatedResultStatus
    duration_ms: Optional[int]
    error_message: Optional[str]
    console_log: Optional[str]
    screenshots: Optional[list] = []
    executed_at: Optional[datetime]
    model_config = {"from_attributes": True}

File: qa-portal/frontend/lib/core/models/automated.dart
class AutomatedSuiteSummary {
  final String id;
  final String projectId;
  final String name;
  final String? description;
  final String? createdAt;
  final int testCount;
  final String? lastRunStatus;
  final String? lastRunAt;
  final int lastRunPassed;
  final int lastRunTotal;

  const AutomatedSuiteSummary({
    required this.id,
    required this.projectId,
    required this.name,
    this.description,
    this.createdAt,
    this.testCount = 0,
    this.lastRunStatus,
    this.lastRunAt,
    this.lastRunPassed = 0,
    this.lastRunTotal = 0,
  });

  factory AutomatedSuiteSummary.fromJson(Map<String, dynamic> json) {
    return AutomatedSuiteSummary(
      id: json['id'] as String,
      projectId: json['project_id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      createdAt: json['created_at'] as String?,
      testCount: json['test_count'] as int? ?? 0,
      lastRunStatus: json['last_run_status'] as String?,
      lastRunAt: json['last_run_at'] as String?,
      lastRunPassed: json['last_run_passed'] as int? ?? 0,
      lastRunTotal: json['last_run_total'] as int? ?? 0,
    );
  }
}

class AutomatedTestCaseModel {
  final String id;
  final String suiteId;
  final String name;
  final String? description;
  final String? sourceTestCaseId;
  final String? scriptCode;
  final String? targetUrl;
  final int order;
  final bool isActive;
  final String? createdAt;

  const AutomatedTestCaseModel({
    required this.id,
    required this.suiteId,
    required this.name,
    this.description,
    this.sourceTestCaseId,
    this.scriptCode,
    this.targetUrl,
    this.order = 0,
    this.isActive = true,
    this.createdAt,
  });

  factory AutomatedTestCaseModel.fromJson(Map<String, dynamic> json) {
    return AutomatedTestCaseModel(
      id: json['id'] as String,
      suiteId: json['suite_id'] as String,
      name: json['name'] as String,
      description: json['description'] as String?,
      sourceTestCaseId: json['source_test_case_id'] as String?,
      scriptCode: json['script_code'] as String?,
      targetUrl: json['target_url'] as String?,
      order: json['order'] as int? ?? 0,
      isActive: json['is_active'] as bool? ?? true,
      createdAt: json['created_at'] as String?,
    );
  }
}

class AutomatedRunSummary {
  final String id;
  final String suiteId;
  final String status;
  final String? environment;
  final String? version;
  final String? startedAt;
  final String? finishedAt;
  final int total;
  final int passed;
  final int failed;
  final int error;
  final int skipped;

  const AutomatedRunSummary({
    required this.id,
    required this.suiteId,
    required this.status,
    this.environment,
    this.version,
    this.startedAt,
    this.finishedAt,
    this.total = 0,
    this.passed = 0,
    this.failed = 0,
    this.error = 0,
    this.skipped = 0,
  });

  factory AutomatedRunSummary.fromJson(Map<String, dynamic> json) {
    return AutomatedRunSummary(
      id: json['id'] as String,
      suiteId: json['suite_id'] as String,
      status: json['status'] as String,
      environment: json['environment'] as String?,
      version: json['version'] as String?,
      startedAt: json['started_at'] as String?,
      finishedAt: json['finished_at'] as String?,
      total: json['total'] as int? ?? 0,
      passed: json['passed'] as int? ?? 0,
      failed: json['failed'] as int? ?? 0,
      error: json['error'] as int? ?? 0,
      skipped: json['skipped'] as int? ?? 0,
    );
  }
}

class AutomatedRunResult {
  final String id;
  final String runId;
  final String automatedTestCaseId;
  final String testName;
  final String status;
  final int? durationMs;
  final String? errorMessage;
  final String? consoleLog;
  final List<String> screenshots;
  final String? executedAt;

  const AutomatedRunResult({
    required this.id,
    required this.runId,
    required this.automatedTestCaseId,
    required this.testName,
    required this.status,
    this.durationMs,
    this.errorMessage,
    this.consoleLog,
    this.screenshots = const [],
    this.executedAt,
  });

  factory AutomatedRunResult.fromJson(Map<String, dynamic> json) {
    return AutomatedRunResult(
      id: json['id'] as String,
      runId: json['run_id'] as String,
      automatedTestCaseId: json['automated_test_case_id'] as String,
      testName: json['test_name'] as String? ?? '',
      status: json['status'] as String,
      durationMs: json['duration_ms'] as int?,
      errorMessage: json['error_message'] as String?,
      consoleLog: json['console_log'] as String?,
      screenshots: (json['screenshots'] as List<dynamic>?)
              ?.map((e) => e.toString())
              .toList() ??
          [],
      executedAt: json['executed_at'] as String?,
    );
  }
}

class RunStatusModel {
  final String id;
  final String status;
  final int total;
  final int completed;
  final int passed;
  final int failed;
  final int error;

  const RunStatusModel({
    required this.id,
    required this.status,
    this.total = 0,
    this.completed = 0,
    this.passed = 0,
    this.failed = 0,
    this.error = 0,
  });

  factory RunStatusModel.fromJson(Map<String, dynamic> json) {
    return RunStatusModel(
      id: json['id'] as String,
      status: json['status'] as String,
      total: json['total'] as int? ?? 0,
      completed: json['completed'] as int? ?? 0,
      passed: json['passed'] as int? ?? 0,
      failed: json['failed'] as int? ?? 0,
      error: json['error'] as int? ?? 0,
    );
  }
}
